File: backend/services/account-service/internal/dto/coffee_bean.go
package dto

import "time"

// 咖啡豆这一半的对外形状。与 fortune_card.go 那份刻意保持同形（流水一行一个结构体、
// 账户一个、查询条件一个），差别来自口径：没有冻结，所以账户上没有 frozen/available 两项；
// 金额是**分**（int64），所以到处是「单位分」这句话要写。

// BeanEntryQuery 是咖啡豆流水查询的过滤条件。零值全部表示「不限」。
//
// 过滤列叫 ReferenceNo 而不是 EntryQuery 里的 OrderNo：豆流水的 reference_no 有三种
// 形状（调整的幂等号、订单号、售后单号），用订单命名会让「按调整号查」变成一件说不出口
// 的事。From/To 是闭开区间 [From, To)，理由与福卡那份逐字相同。
type BeanEntryQuery struct {
	UserID      string
	ReferenceNo string
	EntryType   string
	From        *time.Time
	To          *time.Time
	Page        int
	PageSize    int
}

// BeanAccountResponse 是一个用户的咖啡豆账户。
//
// 没有账户行时余额是 0 而不是 404：账户行是第一次调整或第一次扣减时才懒创建的，一个从没
// 充过豆的用户问自己的余额，答案是「0 分」这件事本身，不是「查不到这个人」。
//
// 没有 frozenBalance / availableBalance 两项（福卡那边有）：豆在支付的那一刻就已经扣走，
// 不存在「冻着但还在」的钱，可用就是余额。补上那两项等于凭空造出一个恒等于余额的数字。
type BeanAccountResponse struct {
	UserID  string `json:"userId"`
	Balance int64  `json:"balance"` // 单位为分
	// HasAccount 区分「有过账户、余额为 0」与「从来没有过账户」——客服看到 0 分时会想知道
	// 是哪一种，两者在流水页上差别很大（一个有历史，一个没有）。
	HasAccount bool      `json:"hasAccount"`
	CreatedAt  time.Time `json:"createdAt"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

// BeanEntryResponse 是咖啡豆流水对外的一行。
//
// 字段名与 admin-web / 小程序读的是同一套（json tag 就是契约）。BalanceAfter 是「变动后
// 余额」而不是当前余额：冲正重放时，只有它能回答「当时是多少」。
type BeanEntryResponse struct {
	ID        string `json:"id"`
	UserID    string `json:"userId"`
	EntryType string `json:"entryType"`
	// Amount 单位为分，**带符号**：充值为正、纠错为负、扣减为负、冲正与它冲掉的那笔相反。
	Amount        int64  `json:"amount"`
	BalanceAfter  int64  `json:"balanceAfter"` // 单位为分
	Title         string `json:"title"`
	ReferenceType string `json:"referenceType"`
	ReferenceID   string `json:"referenceId"`
	ReferenceNo   string `json:"referenceNo"`
	// ReversesEntryID 只有冲正那一条有值：它指向被冲回的那笔扣减。后台客服判断
	// 「这 900 分是退哪一笔的」看的就是它。
	ReversesEntryID string `json:"reversesEntryId"`
	// OperatorID 只有后台调整有值。展示名不在这里：令牌里没有用户名，要按 ID 去身份库解析
	// （见 platform/audit 的 Entry 说明），所以这一层不返回一个恒为空串的 operatorName。
	OperatorID string    `json:"operatorId"`
	Remark     string    `json:"remark"`
	OccurredAt time.Time `json:"occurredAt"`
	CreatedAt  time.Time `json:"createdAt"`
}

// BeanAdjustInput 是后台「余额调整」的请求体。
//
// Amount **带符号**（单位为分）：充值为正、把充错的豆调回来为负。没有单独的「扣回」接口，
// 一个带符号的金额就够了——分成两个接口会让「-100 走哪一个」变成一个新问题，而答案取决于
// 那个接口当时的措辞。
//
// RequestID 是幂等号，由后台每次打开弹窗时生成一个新的：网络抖一下再点一次不会充两次。
// 但同一个 requestId 再来一次**不是**静默回放，而是 409（见 repository.ErrBeanDuplicateRequest）
// ——点这一次的是人，回一句「成功了」会让它再点一次，于是真的充两次。后台拿到 409 应当重新
// 拉余额、先对流水，再决定换不换号重发。
type BeanAdjustInput struct {
	Amount    int64  `json:"amount"`
	RequestID string `json:"requestId"`
	// Remark 是操作人填的理由。它会落进流水的 remark，也是它唯一的去处：
	// admin_operation_logs 没有 reason 列（见 platform/audit 的调用处）。
	Remark string `json:"remark"`
}

// BeanAdjustResponse 是调整之后立刻回给后台的那两个数。
//
// 余额一起给，前端就能把页面上的数字直接改对，不必再发一次读——而且**这个**余额是刚写完
// 那一笔之后的数，重新读一次拿到的可能是别人又动过的余额，与它刚提交的那次调整对不上。
type BeanAdjustResponse struct {
	Balance int64  `json:"balance"` // 单位为分
	EntryID string `json:"entryId"`
}

// MiniappBeansResponse 是小程序端一次拿全的回答：余额 + 一页流水。
//
// 合成一个响应是因为明细页永远两个都要——分成两个接口只会让那一次渲染发两个请求，还得处理
// 「余额是新的、流水是旧的」这种半新半旧的状态。
type MiniappBeansResponse struct {
	Balance int64 `json:"balance"` // 单位为分
	// Entries 只是第一页；Total/Page/PageSize 一起给，滚动加载时不用再问一次「还有没有」。
	Entries  []*BeanEntryResponse `json:"entries"`
	Total    int                  `json:"total"`
	Page     int                  `json:"page"`
	PageSize int                  `json:"pageSize"`
}

File: admin-web/src/services/coffeeBean.ts
import { request } from '@umijs/max';
import type { PageQuery, PageResult } from './pagination';

/**
 * 咖啡豆账户域的接口（account-service）。字段名与后端 dto（internal/dto/coffee_bean.go）
 * 的 json tag 逐字对应，同时也是页面上的 dataIndex——改一处要同批改三处。
 *
 * ⚠️ 这里的「咖啡豆」是**用户维度**的余额（coffee_bean_accounts，键是 user_id），
 * 单位是**分**，与订单金额 1:1。另有一个名字很像的东西：设备维度的
 * `devices.coffee_balance`（在 services/coffeeMachine.ts 里是 listDeviceBalanceEntries），
 * 那是按 device_id 记账的预付余额，两者没有任何关系、
 * 不共用接口、也不共用表名。**读这个文件时不要把两边的金额混着用。**
 *
 * 与福卡那份（fortuneCard.ts）的两处结构差异，都来自口径：
 *   1. 没有冻结——豆在支付那一刻就扣走了，账户上没有 frozen/available 两项。
 *   2. 有写——`adjustCoffeeBeans` 是后台**人工调整**，也全仓唯一能凭空加豆的接口。
 */

// ——— 枚举 ———

/**
 * 流水的账变类型。取值来自 account-service 的 model 常量（BeanEntryType*），与
 * migrations/account 的 CHECK 一致。
 *
 * 加新码而这里没登记时，界面退回显示原始码（同 order.ts / fortuneCard.ts 的规矩）。
 */
export type CoffeeBeanEntryType = 'adjust' | 'consume' | 'reverse';

/** 这条流水为什么发生：后台人工调整 / 订单扣减 / 售后冲正。 */
export type CoffeeBeanReferenceType = 'manual' | 'order' | 'after_sale' | '';

// ——— 形状 ———

/**
 * 一个用户的咖啡豆账户。
 *
 * 没有账户行时余额是 0 且 `hasAccount=false`，**不是 404**——账户行是第一次调整或第一次
 * 扣减时才懒创建的。客服看到 0 分时会想区分「有过账户、花光了」与「从来没有过」，
 * 两者在流水表上差别很大（一个有历史，一个一条都没有）。
 */
export type CoffeeBeanAccount = {
  userId: string;
  /** 单位为分。没有 frozen/available 两项：豆没有冻结态，可用就是余额。 */
  balance: number;
  hasAccount: boolean;
  createdAt: string;
  updatedAt: string;
};

/**
 * 一条咖啡豆流水。
 *
 * `amount` 是**有符号**的、单位为分：调整充值为正、纠错为负、订单扣减为负、冲正与它冲掉的
 * 那笔相反。页面因此不能按「金额」渲染，要带着符号显示，否则一次扣减看起来像又充了一笔。
 */
export type CoffeeBeanEntry = {
  id: string;
  userId: string;
  entryType: CoffeeBeanEntryType;
  /** 单位为分，带符号。 */
  amount: number;
  /**
   * 这笔发生**之后**的余额，不是当前余额。冲正重放时只有它能回答「当时是多少」。
   * 要当前余额请看 getCoffeeBeanAccount 或 BeanAdjustResponse.balance。
   */
  balanceAfter: number;
  /** 给用户看的那一行文案（后端生成，界面原样显示）。 */
  title: string;
  referenceType: CoffeeBeanReferenceType;
  referenceId: string;
  /**
   * 人可读的号。三种形状：调整是幂等号（`adjust:{requestId}` 里的后半段）、扣减是订单号、
   * 冲正是售后单号。所以接口的筛选参数叫 referenceNo 而不是 orderNo。
   */
  referenceNo: string;
  /** 冲正时指向被冲的那笔扣减；其余为空串。 */
  reversesEntryId: string;
  /** 只有后台调整那一条有值：谁动的。是管理员的 user id，不是名字。 */
  operatorId: string;
  /** 操作人填的理由。它也是调整理由唯一的去处（审计表没有 reason 列）。 */
  remark: string;
  /** 业务发生时间（扣减是支付时间、冲正是审核时间），与 createdAt 可能不同。 */
  occurredAt: string;
  createdAt: string;
};

/** 「余额调整」的请求体。 */
export type CoffeeBeanAdjustInput = {
  /** 单位为分，**带符号**：充值为正，把充错的豆调回来为负。0 会被后端拒掉。 */
  amount: number;
  /** 幂等号。每次打开弹窗生成一个新的，见 adjustCoffeeBeans 的注释。 */
  requestId: string;
  remark?: string;
};

/**
 * 调整之后立刻回来的那两个数。
 *
 * 余额一起给，页面就能把数字直接改对，不必再发一次读——而且**这个**余额是刚写完那笔之后的
 * 数，重新读一次拿到的可能是别人又动过的余额。
 */
export type CoffeeBeanAdjustResult = {
  /** 单位为分。 */
  balance: number;
  entryId: string;
};

// ——— 查询参数 ———

/**
 * 流水筛选。`referenceNo` 是精确匹配（后端 BeanEntryQuery），所以调用前 trim。
 *
 * `from`/`to` 是**闭开区间** [from, to)：同一条流水只会落在相邻两天的其中一边，
 * 拼接时不会漏也不会重。必须传带时区的 RFC3339（用 toRFC3339 转，与订单列表同款）。
 */
export type CoffeeBeanEntryQuery = PageQuery & {
  userId?: string;
  referenceNo?: string;
  entryType?: CoffeeBeanEntryType;
  from?: string;
  to?: string;
};

// ——— 接口 ———

/**
 * 查某个用户的咖啡豆账户（客服最常问的那一句：他还有多少豆）。
 *
 * 后端把 `/v1/admin/coffee-beans/entries` 注册在 `/v1/admin/coffee-beans/{userId}`
 * **之前**，所以这里不会被打成「userId 是 entries」——两者在网关与路由上都分得开。
 */
export async function getCoffeeBeanAccount(userId: string) {
  return request<CoffeeBeanAccount>(`/api/v1/admin/coffee-beans/${userId}`);
}

/**
 * 按条件查咖啡豆流水（后台）。小程序用户详情的「咖啡豆账户」区块按 userId 用它。
 */
export async function listCoffeeBeanEntries(params?: CoffeeBeanEntryQuery) {
  return request<PageResult<CoffeeBeanEntry>>('/api/v1/admin/coffee-beans/entries', {
    params,
  });
}

/**
 * 后台人工调整某人的咖啡豆余额。返回调整后的余额，不必再查一次。
 *
 * 权限码是 `account:manage`（不是读那条 `account:read`）：这是**直接加钱**的能力，
 * 充值就是正数，没有审批也没有额度上限，所以读写刻意分开。
 *
 * 同一个 requestId 重复提交时后端回 409（而不是静默回放、也不是假装成功）：上一次多半
 * 已经记过账，只是响应没收到。点这个按钮的是人，回一句「成功了」会让他再点一次，于是
 * 真的充两次。所以调用方必须把 409 和普通失败分开处理，且 409 之后要换一枚新的 requestId
 * ——详情页里的做法见 pages/miniapp-users/index.tsx。
 */
export async function adjustCoffeeBeans(userId: string, data: CoffeeBeanAdjustInput) {
  return request<CoffeeBeanAdjustResult>(
    `/api/v1/admin/coffee-beans/${userId}/adjustments`,
    { method: 'POST', data },
  );
}
